tests/e2e/websocket/common.hpp

```diff
@@ -0,0 +1,309 @@
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#include <algorithm>
+#include <chrono>
+#include <cstdint>
+#include <iostream>
+#include <memory>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <thread>
+#include <utility>
+
+#include <gflags/gflags.h>
+#include <spdlog/spdlog.h>
+#include <unistd.h>
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/asio/strand.hpp>
+#include <boost/beast/core.hpp>
+#include <boost/beast/core/buffers_to_string.hpp>
+#include <boost/beast/ssl/ssl_stream.hpp>
+#include <boost/beast/websocket.hpp>
+#include <json/json.hpp>
+#include <mgclient.hpp>
+
+#include "utils/logging.hpp"
+
+namespace beast = boost::beast;
+namespace http = beast::http;
+namespace websocket = beast::websocket;
+namespace net = boost::asio;
+using tcp = boost::asio::ip::tcp;
+namespace ssl = boost::asio::ssl;
+
+constexpr std::array kSupportedLevels{"debug", "trace", "info", "warning", "error", "critical"};
+
+struct Credentials {
+  std::string_view username;
+  std::string_view passsword;
+};
+
+inline void Fail(beast::error_code ec, char const *what) { std::cerr << what << ": " << ec.message() << "\n"; }
+
+std::string GetAuthenticationJSON(const Credentials &creds) {
+  nlohmann::json json_creds;
+  json_creds["username"] = creds.username;
+  json_creds["password"] = creds.passsword;
+  return json_creds.dump();
+}
+
+template <bool ssl = false>
+class Session : public std::enable_shared_from_this<Session<ssl>> {
+  using std::enable_shared_from_this<Session<ssl>>::shared_from_this;
+
+ public:
+  explicit Session(net::io_context &ioc, ssl::context &ctx, std::vector<std::string> &expected_messages) requires(ssl)
+      : resolver_(net::make_strand(ioc)), ws_(net::make_strand(ioc), ctx), received_messages_{expected_messages} {}
+
+  explicit Session(net::io_context &ioc, std::vector<std::string> &expected_messages) requires(!ssl)
+      : resolver_(net::make_strand(ioc)), ws_(net::make_strand(ioc)), received_messages_{expected_messages} {}
+
+  explicit Session(net::io_context &ioc, ssl::context &ctx, std::vector<std::string> &expected_messages,
+                   Credentials creds) requires(ssl)
+      : resolver_(net::make_strand(ioc)),
+        ws_(net::make_strand(ioc), ctx),
+        received_messages_{expected_messages},
+        creds_{creds} {}
+
+  explicit Session(net::io_context &ioc, std::vector<std::string> &expected_messages, Credentials creds) requires(!ssl)
+      : resolver_(net::make_strand(ioc)),
+        ws_(net::make_strand(ioc)),
+        received_messages_{expected_messages},
+        creds_{creds} {}
```

Comment:
These 2 constructors can be replaced by a single one using Parameter pack: https://en.cppreference.com/w/cpp/language/parameter_pack

I think it's a fun exercise to try to write it yourself. :smile:

Reply:
delegate is also what I had in mind actually (delegate + pack) but @jbajic will figure it out :smile: